File: src/backend/controllers/courseController.js

/**
 * Course controller
 */

const Course = require('../models/Course');

// Create new course
const createCourse = async (req, res) => {
  try {
    // Check if user is a teacher
    if (req.user.role !== 'teacher') {
      return res.status(403).json({ message: 'Only teachers can create courses' });
    }

    const { title, description, category, image, duration, featured } = req.body;

    // Validate input
    if (!title || !description || !category) {
      return res.status(400).json({ message: 'Please provide all required fields' });
    }

    // Create course
    const course = await Course.create({
      title,
      description,
      teacher_id: req.user.id,
      category,
      image,
      duration,
      featured: featured || false
    });

    res.status(201).json({
      success: true,
      course
    });
  } catch (error) {
    console.error('Create course error:', error.message);
    res.status(500).json({ message: 'Server error during course creation' });
  }
};

// Get all courses with filters
const getAllCourses = async (req, res) => {
  try {
    const { category, featured, teacher_id } = req.query;
    
    const filters = {};
    
    if (category) filters.category = category;
    if (featured) filters.featured = featured === 'true';
    if (teacher_id) filters.teacher_id = teacher_id;

    const courses = await Course.getAll(filters);

    res.json({
      success: true,
      count: courses.length,
      courses
    });
  } catch (error) {
    console.error('Get all courses error:', error.message);
    res.status(500).json({ message: 'Server error while retrieving courses' });
  }
};

// Get course by ID
const getCourseById = async (req, res) => {
  try {
    const { id } = req.params;
    
    const course = await Course.findById(id);

    if (!course) {
      return res.status(404).json({ message: 'Course not found' });
    }

    res.json({
      success: true,
      course
    });
  } catch (error) {
    console.error('Get course by ID error:', error.message);
    res.status(500).json({ message: 'Server error while retrieving course' });
  }
};

// Update course
const updateCourse = async (req, res) => {
  try {
    const { id } = req.params;
    
    // Get the course
    const course = await Course.findById(id);

    if (!course) {
      return res.status(404).json({ message: 'Course not found' });
    }

    // Check if user is the teacher of this course
    if (course.teacher_id !== req.user.id && req.user.role !== 'admin') {
      return res.status(403).json({ message: 'Not authorized to update this course' });
    }

    const { title, description, category, image, duration, featured } = req.body;

    // Update course
    const updatedCourse = await Course.update(id, {
      title: title || course.title,
      description: description || course.description,
      category: category || course.category,
      image: image || course.image,
      duration: duration || course.duration,
      featured: featured !== undefined ? featured : course.featured
    });

    res.json({
      success: true,
      course: updatedCourse
    });
  } catch (error) {
    console.error('Update course error:', error.message);
    res.status(500).json({ message: 'Server error during course update' });
  }
};

// Delete course
const deleteCourse = async (req, res) => {
  try {
    const { id } = req.params;
    
    // Get the course
    const course = await Course.findById(id);

    if (!course) {
      return res.status(404).json({ message: 'Course not found' });
    }

    // Check if user is the teacher of this course or an admin
    if (course.teacher_id !== req.user.id && req.user.role !== 'admin') {
      return res.status(403).json({ message: 'Not authorized to delete this course' });
    }

    // Delete course
    await Course.delete(id);

    res.json({
      success: true,
      message: 'Course deleted successfully'
    });
  } catch (error) {
    console.error('Delete course error:', error.message);
    res.status(500).json({ message: 'Server error during course deletion' });
  }
};

// Search courses
const searchCourses = async (req, res) => {
  try {
    const { term } = req.query;
    
    if (!term) {
      return res.status(400).json({ message: 'Search term is required' });
    }

    const courses = await Course.search(term);

    res.json({
      success: true,
      count: courses.length,
      courses
    });
  } catch (error) {
    console.error('Search courses error:', error.message);
    res.status(500).json({ message: 'Server error during course search' });
  }
};

// Get teacher's courses
const getTeacherCourses = async (req, res) => {
  try {
    const courses = await Course.getByTeacher(req.user.id);

    res.json({
      success: true,
      count: courses.length,
      courses
    });
  } catch (error) {
    console.error('Get teacher courses error:', error.message);
    res.status(500).json({ message: 'Server error while retrieving teacher courses' });
  }
};

module.exports = {
  createCourse,
  getAllCourses,
  getCourseById,
  updateCourse,
  deleteCourse,
  searchCourses,
  getTeacherCourses
};

File: src/backend/controllers/course_controller.go
package controllers

import (
	"encoding/json"
	"errors"
	"log"
	"net/http"

	"github.com/nnn20040/shabytdiplomwork/src/backend/models"
	"github.com/nnn20040/shabytdiplomwork/src/backend/repository"

	"github.com/gorilla/mux"
)

func CreateCourse(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value(models.UserContextKey).(string)

	user, err := repository.GetUserByID(r.Context(), userID)
	if err != nil {
		http.Error(w, "user doesnt exists", http.StatusInternalServerError)
		return
	}

	if user.Role != "teacher" && user.Role != "admin" {
		http.Error(w, "Only teachers can create courses", http.StatusForbidden)
		return
	}

	var req models.CreateCourseRequest
	err = json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		http.Error(w, "Invalid request data", http.StatusBadRequest)
		return
	}

	if req.Title == "" || req.Description == "" || req.Category == "" {
		http.Error(w, "Please provide all required fields", http.StatusBadRequest)
		return
	}

	course, err := repository.CreateCourse(r.Context(), user.ID, req)
	if err != nil {
		log.Printf("Create course error: %v", err)
		http.Error(w, "Server error during course creation", http.StatusInternalServerError)
		return
	}

	response := models.Response{
		Success: true,
		Data:    course,
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(response)
}

func GetAllCourses(w http.ResponseWriter, r *http.Request) {
	category := r.URL.Query().Get("category")

	featuredStr := r.URL.Query().Get("featured")
	featured := false
	featuredSpecified := false
	if featuredStr != "" {
		featured = featuredStr == "true"
		featuredSpecified = true
	}

	teacherID := r.URL.Query().Get("teacher_id")

	courses, err := repository.GetAllCourses(r.Context(), category, teacherID, featured, featuredSpecified)
	if err != nil {
		log.Printf("Get all courses error: %v", err)
		http.Error(w, "Server error while retrieving courses", http.StatusInternalServerError)
		return
	}

	response := models.Response{
		Success: true,
		Data: map[string]interface{}{
			"count":   len(courses),
			"courses": courses,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func GetCourseByID(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	id := vars["id"]

	course, err := repository.GetCourseByID(r.Context(), id)
	if err != nil {
		if errors.Is(err, models.ErrNotFound) {
			http.Error(w, "Course not found", http.StatusNotFound)
		} else {
			log.Printf("Get course by ID error: %v", err)
			http.Error(w, "Server error while retrieving course", http.StatusInternalServerError)
		}
		return
	}

	response := models.Response{
		Success: true,
		Data: map[string]interface{}{
			"course": course,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func UpdateCourse(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value(models.UserContextKey).(string)

	user, err := repository.GetUserByID(r.Context(), userID)
	if err != nil {
		http.Error(w, "user doesnt exists", http.StatusInternalServerError)
		return
	}

	vars := mux.Vars(r)
	id := vars["id"]

	course, err := repository.GetCourseByID(r.Context(), id)
	if err != nil {
		if err == models.ErrNotFound {
			http.Error(w, "Course not found", http.StatusNotFound)
		} else {
			log.Printf("Get course by ID error: %v", err)
			http.Error(w, "Server error while retrieving course", http.StatusInternalServerError)
		}
		return
	}

	if course.TeacherID != user.ID && user.Role != "admin" {
		http.Error(w, "Not authorized to update this course", http.StatusForbidden)
		return
	}

	var req models.UpdateCourseRequest
	err = json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		http.Error(w, "Invalid request data", http.StatusBadRequest)
		return
	}

	if req.Title == "" {
		req.Title = course.Title
	}
	if req.Description == "" {
		req.Description = course.Description
	}
	if req.Category == "" {
		req.Category = course.Category
	}
	if req.Image == "" {
		req.Image = course.Image
	}
	if req.Duration == "" {
		req.Duration = course.Duration
	}

	updatedCourse, err := repository.UpdateCourse(r.Context(), id, req)
	if err != nil {
		log.Printf("Update course error: %v", err)
		http.Error(w, "Server error during course update", http.StatusInternalServerError)
		return
	}

	response := models.Response{
		Success: true,
		Data: map[string]interface{}{
			"course": updatedCourse,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func DeleteCourse(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value(models.UserContextKey).(string)

	user, err := repository.GetUserByID(r.Context(), userID)
	if err != nil {
		http.Error(w, "user doesnt exists", http.StatusInternalServerError)
		return
	}

	vars := mux.Vars(r)
	id := vars["id"]

	course, err := repository.GetCourseByID(r.Context(), id)
	if err != nil {
		if err == models.ErrNotFound {
			http.Error(w, "Course not found", http.StatusNotFound)
		} else {
			log.Printf("Get course by ID error: %v", err)
			http.Error(w, "Server error while retrieving course", http.StatusInternalServerError)
		}
		return
	}

	if course.TeacherID != user.ID && user.Role != "admin" {
		http.Error(w, "Not authorized to delete this course", http.StatusForbidden)
		return
	}

	err = repository.DeleteCourse(r.Context(), id)
	if err != nil {
		log.Printf("Delete course error: %v", err)
		http.Error(w, "Server error during course deletion", http.StatusInternalServerError)
		return
	}

	response := models.Response{
		Success: true,
		Message: "Course deleted successfully",
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func SearchCourses(w http.ResponseWriter, r *http.Request) {
	term := r.URL.Query().Get("term")
	if term == "" {
		http.Error(w, "Search term is required", http.StatusBadRequest)
		return
	}

	courses, err := repository.SearchCourses(r.Context(), term)
	if err != nil {
		log.Printf("Search courses error: %v", err)
		http.Error(w, "Server error during course search", http.StatusInternalServerError)
		return
	}

	response := models.Response{
		Success: true,
		Data: map[string]interface{}{
			"count":   len(courses),
			"courses": courses,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func GetTeacherCourses(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value(models.UserContextKey).(string)

	user, err := repository.GetUserByID(r.Context(), userID)
	if err != nil {
		http.Error(w, "user doesnt exists", http.StatusInternalServerError)
		return
	}

	if user.Role != "teacher" && user.Role != "admin" {
		http.Error(w, "Only teachers can access their courses", http.StatusForbidden)
		return
	}

	courses, err := repository.GetCoursesByTeacher(r.Context(), user.ID)
	if err != nil {
		log.Printf("Get teacher courses error: %v", err)
		http.Error(w, "Server error while retrieving teacher courses", http.StatusInternalServerError)
		return
	}

	response := models.Response{
		Success: true,
		Data: map[string]interface{}{
			"count":   len(courses),
			"courses": courses,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}
